Feature/up low case btns

sviat-barbutsa opened the pull request:
- Implemented text-transform functionality with the Button component. Added tests of this functionality.

sviat-barbutsa committed:
- removed mousePositionView
- removed stores
- added appModel to the MobxProvider to make it injectable
- refactored AboutPage as we don't have the comboboxStore anymore. Later I'll add proper features to the appModel
- added missed properties to the IAppModelInterface
- added comments etc. why some changes had been made
- implemented text input UI and logic
- added combobox Item etc. to the appModel …
- refactored ComboboxItem to work in the new flow
- added a simple mapper which maps the inputSource to options
- created a dedicated SelectedItem type and placed it in the Combobox definition file
- renamed the Props type to ComboboxProps
- changed type of the inputSource. Added comments about it
- implemented a Combobox view in the Home component

sviat-barbutsa committed:
- implemented a Button component
- added a needed functionality to the appModel
- wrote tests for text transform functionality

nebosite commented on src/models/appmodel.ts:
```diff
@@ -13,7 +13,23 @@ export class AppModel implements IAppModel {
     set textInput(value: string) { this._textInput = value; }
+    get textInputLength(): number { return this._textInput.length; }
+
+    @observable private _transformationText = "Here is some dummy text.";
```
The buttons should act on _textInput.   Remove _transformationText

sviat-barbutsa replied:
Done
Attacked the demo gif to the Trello card

nebosite commented on src/models/__tests__/appmodel.test.ts:
```diff
@@ -0,0 +1,25 @@
+
+	it(`should make the transformation text lowercased
+		 when setLowercase method called`, () => {
+			 const initialText = appModel.transformationText;
```
Tests need to be explicit:
appModel.inputText = "HI THERE";
expect(appModel.inputText).toBe("HI THERE");
appModel.setLowerCase();
expect(appModel.inputText).toBe("hi there");

sviat-barbutsa replied:
Done

nebosite commented on src/models/__tests__/appmodel.test.ts:
```diff
@@ -0,0 +1,25 @@
+
+	it(`should make the transformation text uppercased
+		 when setUppercase method called`, () => {
+			 const initialText = appModel.transformationText;
```
Make the test explicit

sviat-barbutsa replied:
Done

nebosite commented on src/models/appmodel.ts:
```diff
@@ -1,6 +1,6 @@
-import { IAppModel } from "./i_appmodel"
-
+import { IAppModel } from "./i_appmodel";
+import { ComboboxItem } from "shared/components/Combobox";
```
Remove this line.  The appmodel must never know about ComboboxItem- this breaks the MVVM abstraction.

sviat-barbutsa replied:
It's a Type. We aren't using the `types` folder so I put this type to the Combobox component.
I need this type to mark that some property in this model will have a ComboboxItem type

We can add a Suffix Type after name eg. ComboboxItemType etc.

in case of the `types` folder, we wouldn't import this type

nebosite replied:
OK, leave it in and I will show what I mean during refactoring after this PR.

sviat-barbutsa replied:
Sorry Erick, I really don't understand how can I change it without `types` folder.
In TS as like in C# I need to assign to the variable\property some type.
This type is defined in Combobox definition file (like you asked me before) and to let the Compiler know that i_appmodel property selectedItem should be a type of SelectedItem I need to import it directly.
Here is the screen
![image](https://user-images.githubusercontent.com/9897782/69911854-4a8a9700-142a-11ea-9837-4c72438bc3e1.png)


Looking forward to see your solution

nebosite replied:
It's OK.  I'm ready to play with the code at this point and it will be easier to show what I mean by making my own changes.  It's also quite possible I'm missing something and I'll learn it by actually trying to change the code.

nebosite commented on src/models/appmodel.ts:
```diff
@@ -13,7 +13,23 @@ export class AppModel implements IAppModel {
     set selectedFlavorIndex(value: number) { this._selectedFlavorIndex = value; }
 
+    // we can create a dedicated Class for SelectedItem like: class SelectedItem etc. but it will be too complicated
+    @observable private _selectedItem = { label: "", value: "" };
```
_selectedItem must be the same type as the items source array.

sviat-barbutsa replied:
I see. WIll extract the value string from the Select object.

As the ReactSelect passes the label and value object on Select and we need to map string[] before passing options to the ReactSelect and the same when we are selecting item and making it _selectedItem value, I think, in this case we can create some mapper regarding this reason.

sviat-barbutsa replied:
Done, added a simple mapper

nebosite commented on src/models/i_appmodel.ts:
```diff
@@ -1,3 +1,22 @@
+import { ComboboxItem } from "shared/components/Combobox";
```
Remove this

sviat-barbutsa replied:
The same stuff about Type

nebosite replied:
OK, I'll make the change when I simplify the code.

nebosite commented on src/index.tsx:
```diff
@@ -5,15 +5,19 @@ import { BrowserRouter as Router } from "react-router-dom";
+// TODO: Delete after reading. We could do it in the appmodel file. + your class namings etc. a bit strange.
+// Usually Classes file names has PaskalCase format and it let's us know that it's a Class or Component. 
+// if it's a singleton and it exports the instance of itself so we can use camelCase
+// it was FYI
```
Feel freem to rename to pascal case

nebosite commented on src/models/i_appmodel.ts:
```diff
@@ -1,3 +1,22 @@
+ You can write this interface in the src/typings folder and all types will be dynamically fetched from there.
+ You won't need anymore to directly import this time into files that uses it.
+
+ The direct import could be useful if you want to quick jump to the source file via ctrl+click on the type
```
Let's leave the direct imports for clarity

nebosite commented on src/models/i_appmodel.ts:
```diff
@@ -1,3 +1,22 @@
+    flavors: string[];
+    textInput: string;
+    textInputLength: number;
+    selectedFlavorIndex: number;
```
If we implement selected item, we can probably remove this

nebosite commented on src/shared/components/Button.tsx:
```diff
@@ -0,0 +1,15 @@
+import * as React from "react";
```
Do we really need a separate button component?  Can this simple be a tag on the home page?

nebosite commented on src/shared/components/Combobox.tsx:
```diff
@@ -5,6 +5,11 @@ import CreatableSelect from "react-select/creatable";
 // and the implementation wasn't designed to be testable from outside
 
+export type ComboboxItem = {
+	value: string;
```
value should be of type any

velidan replied:
Np, but could you, please, explain why do you think it'd be better?
 if it should be string only because ReactSelect expects to receive exactly that type.
In case of own component, we could handle any types etc.
![image](https://user-images.githubusercontent.com/9897782/69912059-21b7d100-142d-11ea-95c6-53afe96be810.png)

sviat-barbutsa replied:
done

nebosite commented on src/shared/components/Combobox.tsx:
```diff
@@ -22,23 +27,25 @@ enum OnChangeEnum {
-  options: ComboboxItem[];
+type ComboboxProps = {
+  // but we need to know what data type handles our Combobox otherwise it will fall in case of incorrect data like number[] or object[]
+  // maybe string[] if we know that the combobox should handle this an array of string?
```
There are a few techniques that can be used to get the string representation of objects.

The default way to get the label would be to get the tostring result, which works on any object.
Add a property to the combobox called "displayPath" that allows the caller to specify the property to use as a lable.
Add a property called "labelFunction" which is a lambda provided by the caller to convert the object into a text label

sviat-barbutsa replied:
You are right, but in that case, we need to implement them.
So if we start to use non-string objects we need to add our custom string representation and if we will use some specific methods like toLowerCase (I used to it to get the ID from the label before I changed a possible type to ANY. Now it's removed)

nebosite replied:
In typescript, all value types have a native string representation that can be obtained from object.toString().    Complex type can override .toString to provide a string representation.  The comboxBox needs no special knowledge about the types.  It can simply call .toString on the objects in the itemsSource to get the labels.

nebosite commented on src/shared/components/Combobox.tsx:
```diff
@@ -49,9 +56,15 @@ export default function Combobox(props: Props) {
+     Also in case of complicated data this mapper should be placed in the different file
+     in the mapper folder and it should map source to the Entities. like new SelectedItem() etc.
+  */
+  const options = itemsSource.map(itemName => ({ value: itemName.toLowerCase(), label: itemName }));
```
I think this will be simpler if the options is simple an array of combobox items where the value points to the original object.  It is possible that two items might have the same name and we don't want that to cause a collision.

sviat-barbutsa replied:
hmmm, but what is the reason to let the users choose one from two the same options in the combobox? I don't understand. I thought that in the combobox should be only Unique items. 
Btw, we definitely need some mapper because ReactSelect works with that type: { label : <val>, value : <val> }
![image](https://user-images.githubusercontent.com/9897782/69912210-a2c39800-142e-11ea-98e5-c86f52aacb17.png)

nebosite replied:
General purpose comboboxes cannot assume that items will have unique string representations.  (Well they could, but in practice this creates more problems than it solves.  e.g.:  A programmer could forget to implement .toString to objects dynamically added to a combobox.   The user would not like to see a crash in this case. )

nebosite commented on src/views/HomePage.tsx:
```diff
@@ -42,20 +46,70 @@ export default class Home extends React.Component<HomeProperties> {
+          {this.props.appModel.mousePosition.x}, 
+          {this.props.appModel.mousePosition.y}</h3>
+
+        <br />
```
Maybe just an <hr> here to make the code have fewer lines

sviat-barbutsa replied:
done

nebosite commented on src/views/HomePage.tsx:
```diff
@@ -42,20 +46,70 @@ export default class Home extends React.Component<HomeProperties> {
+        <Button onClick={appModel.setLowercase} >Lower Case</Button>
+        <br />
+        <br />
+        Text that will be transformed: <b>{ appModel.transformationText }</b>
```
Move the buttons up and have them apply to the text input

sviat-barbutsa replied:
done

nebosite reviewed (changes requested):
It looks like this PR is a superset of the other, so I will leave all my comments on this PR.
I like the work that is happening here.  There are some important changes to make to keep this pure MVVM.  
Let's focus on getting this PR checked in before moving on to new tasks.

sviat-barbutsa committed:
- changed ComboboxItem value type to any. Added label toString convertation
- refactored appmodel. Removed transformation text and let the upper|lower methods to work with the textInput value
- removed <br /> from the UI, added inline styles to <hr /> to keep normal space format
- made tests a bit more explicit
- implemented a simple mapper for the Combobox Source Items
